File: src/lib/planLimits.ts
export type Plan = 'free' | 'pro' | 'studio';
export type QuotaKey =
  | 'fullSongs'
  | 'beats'
  | 'clips30s'
  | 'images'
  | 'lyrics'
  | 'audioEnhance'
  | 'audioEnhanceStandard';

export const UNLIMITED = 999999;

export const MONTHLY_LIMITS: Record<Plan, Record<QuotaKey, number>> = {
  free: {
    fullSongs:            2,
    beats:                30,
    clips30s:             20,
    images:               30,
    lyrics:               UNLIMITED,
    audioEnhance:         0,
    audioEnhanceStandard: 15,
  },
  pro: {
    fullSongs:            30,
    beats:                UNLIMITED,
    clips30s:             UNLIMITED,
    images:               UNLIMITED,
    lyrics:               UNLIMITED,
    audioEnhance:         UNLIMITED,
    audioEnhanceStandard: UNLIMITED,
  },
  studio: {
    fullSongs:            UNLIMITED,
    beats:                UNLIMITED,
    clips30s:             UNLIMITED,
    images:               UNLIMITED,
    lyrics:               UNLIMITED,
    audioEnhance:         UNLIMITED,
    audioEnhanceStandard: UNLIMITED,
  },
};

export const DAILY_LIMITS: Record<Plan, Partial<Record<QuotaKey, number>>> = {
  free: {
    beats:                5,
    clips30s:             3,
    images:               5,
    lyrics:               10,
    audioEnhanceStandard: 3,
  },
  pro: {},
  studio: {},
};

export const PLAN_PRICES = {
  free:   { monthly: 0,   annual: 0,   label: 'Free'   },
  pro:    { monthly: 19,  annual: 190, label: 'Pro'    },
  studio: { monthly: 49,  annual: 490, label: 'Studio' },
} as const;

export const ONCE_OFF_PRICES = {
  audioEnhance: { amountZAR: 100, label: 'Single audio enhancement' },
} as const;

export function isUnlimited(val: number): boolean {
  return val >= UNLIMITED;
}

export function isAtMonthlyLimit(plan: Plan, key: QuotaKey, used: number): boolean {
  const limit = MONTHLY_LIMITS[plan][key];
  return !isUnlimited(limit) && used >= limit;
}

export function isAtDailyLimit(plan: Plan, key: QuotaKey, used: number): boolean {
  const limit = DAILY_LIMITS[plan]?.[key];
  if (limit === undefined) return false;
  return used >= limit;
}

export function monthlyRemaining(plan: Plan, key: QuotaKey, used: number): number | null {
  const limit = MONTHLY_LIMITS[plan][key];
  if (isUnlimited(limit)) return null;
  return Math.max(0, limit - used);
}

export function dailyRemaining(plan: Plan, key: QuotaKey, used: number): number | null {
  const limit = DAILY_LIMITS[plan]?.[key];
  if (limit === undefined) return null;
  return Math.max(0, limit - used);
}

File: backend/core/plan_limits.py
from enum import Enum
from typing import Union

class Plan(str, Enum):
    FREE   = 'free'
    PRO    = 'pro'
    STUDIO = 'studio'

INF = float('inf')

MONTHLY_LIMITS: dict[str, dict[str, float]] = {
    'free':   { 'full_songs': 2,   'beats': 30,  'clips_30s': 20, 'images': 30, 'lyrics': INF, 'audio_enhance': 0,   'audio_enhance_standard': 15 },
    'pro':    { 'full_songs': 30,  'beats': INF, 'clips_30s': INF,'images': INF,'lyrics': INF, 'audio_enhance': INF, 'audio_enhance_standard': INF },
    'studio': { 'full_songs': INF, 'beats': INF, 'clips_30s': INF,'images': INF,'lyrics': INF, 'audio_enhance': INF, 'audio_enhance_standard': INF },
}

DAILY_LIMITS: dict[str, dict[str, float]] = {
    'free':   { 'beats': 5, 'clips_30s': 3, 'images': 5, 'lyrics': 10, 'audio_enhance_standard': 3 },
    'pro':    { 'beats': INF, 'clips_30s': INF, 'images': INF, 'lyrics': INF, 'audio_enhance_standard': INF },
    'studio': { 'beats': INF, 'clips_30s': INF, 'images': INF, 'lyrics': INF, 'audio_enhance_standard': INF },
}

def is_at_limit(plan: str, key: str, used: int, limit_type: str = 'monthly') -> bool:
    limits = MONTHLY_LIMITS if limit_type == 'monthly' else DAILY_LIMITS
    limit = limits.get(plan, {}).get(key, INF)
    return limit != INF and used >= limit

def get_remaining(plan: str, key: str, used: int, limit_type: str = 'monthly') -> float:
    limits = MONTHLY_LIMITS if limit_type == 'monthly' else DAILY_LIMITS
    limit = limits.get(plan, {}).get(key, INF)
    if limit == INF:
        return INF
    return max(0, limit - used)
